lib/pool.js

```diff
@@ -0,0 +1,93 @@
+
+const debug                = require('debug')('kuss:pool')
+const R                    = require('ramda')
+const Bluebird             = require('bluebird')
+const GenericPool          = require('generic-pool')
+const { FibonacciBackoff } = require('simple-backoff')
+
+const CONFIG_PROP_SERVER    = 'server'
+const CONFIG_PROP_VALIDATOR = 'validateAsync'
+const CONFIG_PROP_TEST      = 'testOnBorrow'
+
+const DEFAULT_CONFIG = {
+  max               : 10
+, min               : 2
+, idleTimeoutMillis : 30000
+, log               : false
+, testOnBorrow      : false
+, logger            : console
+}
+
+
+const shutdown = (pool) => () => pool.drain().then(() => pool.clear())
+
+
+const getServerConfig = R.propOr({}, CONFIG_PROP_SERVER)
+
+
+const addBackoff = (factory) => {
+  const backoff   = new FibonacciBackoff()
+  const orgCreate = factory.create
+
+  const run = () => orgCreate().catch( e => new Bluebird((resolve, reject) => {
+    const wait = backoff.next()
+
+    debug('could not create resource - waiting %d ms - %s', wait, e.message)
+
+    setTimeout(() => reject(e), wait)
+  }))
+
+  factory.create = run
+
+  return factory
+
+}
+
+
+const buildFactory = (builder, default_config) => R.compose(
```

Comment:
Why not auto-curry these functions that seem to be arbitrarily tupled/curried? With the way that javascript will allow this first function to take three params and silently discard the third leads to very annoying, difficult to track bugs.